1219 v6 patch job type

JohnPTobe opened the pull request.

dfaller commented on scale/docs/rest/v6/job_type.rst:
```diff
@@ -418,193 +422,195 @@ Response: 200 OK
+      "max_scheduled": 1,
+      "docker_image": "test-1.0.0-seed:1.0.0",
+      "configuration": {
+        "version": "2.0",
```
The API should accept version 6 of the configuration. The only change here is to omit the `version` field.

JohnPTobe replied:
Should it be removed or just changed to 6 as these are POST examples?

dfaller replied:
Just remove the `"version": "2.0",` line, since the rest is still version 6 compliant.

dfaller commented on scale/docs/rest/v6/job_type.rst:
```diff
@@ -418,193 +422,195 @@ Response: 200 OK
+		"icon_code": "f1c5", 
+		"is_active": true, 
+		"is_paused": false, 
+		"is_system": true, 
```
Should flip `is_system` to `false` in doc since that should be its value.

dfaller commented on scale/docs/rest/v6/job_type.rst:
```diff
@@ -613,46 +619,160 @@ Response: 200 OK
+
+**Example POST /v6/job-types/ API call**
+
+Request: POST http://.../v6/job-types/
```
The `/validation/` is missing from the end of the example URLs here.

dfaller commented on scale/docs/rest/v6/job_type.rst:
```diff
@@ -840,579 +792,80 @@ Response: 200 OK
+      "is_paused": false,
+      "max_scheduled": 1,
+      "configuration": {
+        "version": "2.0",
```
Same comment as above about removing `version` from the configuration JSON. This should be done everywhere in the v6 Job Type docs.

dfaller commented on scale/job/models.py:
```diff
@@ -2460,13 +2465,16 @@ def create_job_type_v6(self, manifest_dict, trigger_rule_dict=None, configuratio
 
         secrets = None
         if configuration_dict:
             configuration = JobConfigurationV6(configuration_dict, do_validate=True).get_configuration()
```
I recommend moving this line, `configuration = JobConfigurationV6(configuration_dict, do_validate=True).get_configuration()` into the view (within a try/except for `InvalidConfiguration`) and pass the general configuration object into this method instead of passing the dict.

dfaller commented on scale/job/models.py:
```diff
@@ -2460,13 +2465,16 @@ def create_job_type_v6(self, manifest_dict, trigger_rule_dict=None, configuratio
-        trigger_rule = self._create_seed_job_trigger_rule(manifest, trigger_rule_dict)
+        trigger_rule = None
+        if trigger_rule_dict:
+            trigger_rule = self._create_seed_job_trigger_rule(manifest, trigger_rule_dict)
```
Go ahead and remove trigger_rule_dict from this method. v6 will not support trigger rules.

dfaller replied:
Also remove `**kwargs`. We should only allow specific parameters into this method and set specific fields on the job type model.

dfaller replied:
Also remove `_create_seed_job_trigger_rule()`

dfaller replied:
Also make the same changes to `edit_job_type_v6()`. For editing, remove trigger rule and kwargs, and make sure editable fields like is_active are passed in. Also pass in the configuration object just like `create_job_type_v6()`. Pass in the manifest class as well instead of the dict and make sure the manifest updates fields like title, description etc when editing.

dfaller replied:
How does `create_job_type_v6()` work with existing job types? It seems like it always saves a new one which would error when an existing job type with the same name and version already exists.

EDIT: Never mind, I see you split the logic in the view. That's fine.

dfaller commented on scale/job/models.py:
```diff
@@ -3184,6 +3193,73 @@ def validate_job_type_v5(self, name, version, interface, error_mapping=None, tri
+                pass
+
+        if manifest:
+            try:
```
This validation block checking against recipe types can be removed. My new recipe work now pins recipes to a job type revision, so changing a job type never breaks a recipe now.

dfaller commented on scale/job/views.py:
```diff
@@ -277,57 +278,75 @@ def create_v6(self, request):
+        existing_job_type = JobType.objects.filter(name=name, version=version).first()
+        if not existing_job_type:
+            try:
+                with transaction.atomic():
```
The `with transaction.atomic():` is unneeded here.

dfaller commented on scale/job/views.py:
```diff
@@ -277,57 +278,75 @@ def create_v6(self, request):
+                message = 'Job type configuration invalid'
+                logger.exception(message)
+                raise BadParameter('%s: %s' % (message, unicode(ex)))
+            except InvalidResources as ex:
```
Should be able to remove `InvalidResources` except.

dfaller commented on scale/job/views.py:
```diff
@@ -277,57 +278,75 @@ def create_v6(self, request):
+                                                                  configuration_dict=configuration_dict,
+                                                                  **extra_fields)
+    
+            except (InvalidJobField, InvalidTriggerType, InvalidTriggerRule, InvalidTriggerMissingConfiguration,
```
Remove the `**extra_fields` and remove catching these exceptions, except `InvalidSecretsConfiguration`.

dfaller commented on scale/job/views.py:
```diff
@@ -277,57 +278,75 @@ def create_v6(self, request):
+                with transaction.atomic():
+                    # Edit the job type
+                    JobType.objects.edit_job_type_v6(job_type_id=existing_job_type.id, manifest_dict=manifest_dict,
+                                                     configuration_dict=configuration_dict, **extra_fields)
```
Same changes as comment above for creating a job type.

JohnPTobe commented:
I addressed all the issues from the PR but moving the configuration logic as I feel the logic belongs inside the model and it saves passing the secrets as well as the configuration.

dfaller commented on scale/job/models.py:
```diff
@@ -2435,53 +2435,45 @@ def create_job_type_v5(self, name, version, interface, trigger_rule=None, error_
-            configuration.validate(manifest)
-            secrets = configuration.remove_secret_settings(manifest)
-            configuration_dict = convert_config_to_v6_json(configuration).get_dict()
-            job_type.configuration = configuration_dict
```
Sorry if I wasn't clear. My intent was for the view to have just the line converting the v6 JSON:
`configuration = JobConfigurationV6(configuration_dict, do_validate=True).get_configuration()`

This method would then contain the rest of the code, something like this:
```
if not configuration:
    configuration = JobConfiguration()
configuration.validate(manifest)
secrets = configuration.remove_secret_settings(manifest)
job_type.configuration = convert_config_to_v6_json(configuration).get_dict()

```

This way the view is handling the JSON validation logic with dealing with user input and the manager method is handling business logic/validation and operates on the logic class (`JobConfiguration`).

dfaller commented on scale/job/models.py:
```diff
@@ -2643,6 +2648,11 @@ def edit_job_type_v6(self, job_type_id, manifest=None, docker_image=None, icon_c
             job_type.author_url = manifest.get_maintainer()['url']
             job_type.save()
```
After the job type save here, we need a call to `JobTypeRevision.objects.create_job_type_revision(job_type)` just like during job type creation so there is a new revision for the manifest changes. The call must be after the job type save because the call assumes that `manifest` and `revision_num` have been updated.

dfaller replied:
Never mind, it's there. :)
